src/server/multi_svr.c

```diff
@@ -306,3 +524,339 @@ gen_svr_inst_id(void)
+ * @brief find the number of peer server replies
+ * which needs to be ack'd
+ * 
+ * @return int 
+ */
+int
+num_pending_peersvr_rply(void)
+{
+	server_t *psvr;
+	int ct = 0;
+	svrinfo_t *svr_info;
+
+	for (psvr = GET_NEXT(peersvrl);
+	     psvr; psvr = GET_NEXT(psvr->mi_link)) {
+		svr_info = psvr->mi_data;
+		ct += svr_info->num_pending_rply;
+	}
+
+	return ct;
+}
+
+/**
+ * @brief Go through peer server list and
+ * poke if any of them is down
+ */
+void
+poke_peersvr(void)
+{
+	server_t *psvr;
+
+	for (psvr = GET_NEXT(peersvrl);
+	     psvr; psvr = GET_NEXT(psvr->mi_link)) {
+		connect_to_peersvr(psvr);
+	}
+}
+
+/**
+ * @brief save the resource update to the cache
+ * DECR requests will free the corresponding INCR from cache
+ * 
+ * @param[in,out] pobj - pointer to peer server struct
+ * @param[in,out] ru_new  - resource update which needs to be saved
+ * @return int 
+ * @retval !0 - error code
+ */
+static int
+save_resc_update(void *pobj, psvr_ru_t *ru_new)
+{
+	psvr_ru_t *ru_old = NULL;
+	int rc = 0;
+	server_t *psvr = pobj;
+
+	if (!ru_new || !ru_new->jobid)
+		return -1;
+
+	pbs_idx_find(psvr->mi_rsc_idx, (void **) &ru_new->jobid, (void **) &ru_old, NULL);
+	if (!ru_old && ru_new->op == INCR) {
+		if (!psvr->mi_rsc_idx)
+			psvr->mi_rsc_idx = pbs_idx_create(0, 0);
+		rc = pbs_idx_insert(psvr->mi_rsc_idx, ru_new->jobid, ru_new);
+	} else if (ru_old && ru_new->op == DECR) {
+		pbs_idx_delete(psvr->mi_rsc_idx, ru_old->jobid);
+		delete_clear_link(&ru_old->ru_link);
+		free_ru(ru_old);
+	} else {
+		rc = PBSE_DUPRESC;
+		log_eventf(PBSEVENT_DEBUG, PBS_EVENTCLASS_SERVER, LOG_WARNING, __func__,
+			   "Duplicate resource update received for job %s , op=%d", ru_new->jobid, ru_new->op);
+		delete_clear_link(&ru_new->ru_link);
+		free_ru(ru_new);
+	}
+
+	return rc;
+}
+
+/**
+ * @brief handler for resource update from a peer server
+ * 
+ * @param[in] stream - connection stream
+ * @param[in] ru_head - head of resource update list
+ * @param[in,out] psvr - peer server object
+ */
+void
+req_resc_update(int stream, pbs_list_head *ru_head, void *psvr)
+{
+	attribute pexech;
+	psvr_ru_t *ru_cur;
+	psvr_ru_t *ru_nxt;
+	int op = 0;
+	int rc = 0;
+
+	for (ru_cur = GET_NEXT(*ru_head);
+	     ru_cur; ru_cur = ru_nxt) {
+		ru_nxt = GET_NEXT(ru_cur->ru_link);
+
+		log_eventf(PBSEVENT_DEBUG, PBS_EVENTCLASS_SERVER, LOG_DEBUG, __func__,
+			   "received update jobid=%s, op=%d, execvnode=%s",
+			   ru_cur->jobid, ru_cur->op, ru_cur->execvnode);
+
+		if (ru_cur->op == INCR)
+			op = INCR;
+
+		rc = save_resc_update(psvr, ru_cur);
+		if (rc == PBSE_DUPRESC)
+			continue;
+
+		update_jobs_on_node(ru_cur->jobid, ru_cur->execvnode, ru_cur->op, ru_cur->share_job);
+		job_attr_def[JOB_ATR_exec_vnode].at_decode(&pexech, job_attr_def[JOB_ATR_exec_vnode].at_name,
+							   NULL, ru_cur->execvnode);
+		update_node_rassn(&pexech, ru_cur->op);
+
+		if (ru_cur->op == DECR) {
+			delete_clear_link(&ru_cur->ru_link);
+			free_ru(ru_cur);
+		}
+	}
+
+	if (op == INCR)
```

Comment:
Why ack in case of INCR only? DECR could also create inconsistencies!!!

Reply:
INCR will result in over-consumption and DECR results in under-utilization. But an under-utilization can be filled in the very next scheduling cycle.
This will be similar to an under-utilization scenario in single-server mode. The job has ended and the obit has been sent. But the server processes the sched stat request before processing obit. This should be fine if the scheduler can fill those resources in the next cycle.

Reply:
sure, that is fine. But what I am asking is that is there is a pending ack (due to not acking for DECR), then the sending server will have a "pending_acks" count right? And that would stop scheduling during the fence operation?

Reply:
Pending ACKs are only incremented for INCR updates.

Reply:
ok